src/disk.hpp

```diff
@@ -232,14 +232,23 @@ struct BufferedDisk : Disk
         NeedReadCache();
         // all allocations need 7 bytes head-room, since
         // SliceInt64FromBytes() may overrun by 7 bytes
-        if (read_buffer_start_ <= begin && read_buffer_start_ + read_buffer_size_ > begin + length + 7) {
+        if (read_buffer_start_ <= begin
+            && read_buffer_start_ + read_buffer_size_ >= begin + length
+            && read_buffer_start_ + read_ahead >= begin + length + 7)
```

Comment:
This condition needs to check two separate criteria:

1. The read request is within the cache. i.e. we have all these bytes in the cache
2. The read request has at least 7 bytes of margin before the end of the *allocation*

The existing check conflates these two and makes sure there are at least 7 bytes of *valid* cache after the current request. This causes the last few requests to hit the disk and wipe the cache even though they don't need to.